Send post_id to the REST API in the ServerSideRender component within the Editor

alexsanford opened the pull request:
## Description
<!-- Please describe what you have changed or added -->

Fixes https://github.com/WordPress/gutenberg/issues/8178

This PR adds a wrapper around the `ServerSideRender` component within the `@wordpress/editor` package that adds a `post_id` parameter to the REST API call. This ensures the global `$post` object is set properly. Previously, it was only set when rendering to the frontend, but not when rendering for the API call.

This PR is an iteration of this previous PR: https://github.com/WordPress/gutenberg/pull/8270

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

This was tested locally using a dynamic block that displayed information from the `$post` object, ensuring the post information was rendered properly on the frontend and in the editor using `ServerSideRender`.

An automated test was added to ensure that the parameter is added properly to the URL path.

Also I verified that the current two blocks using `ServerSideRender` (Archives and Latest Comments) still work as expected.

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

This change should be a non-breaking enhancement.

That said, I did change the name of the `rendererPathWithAttributes` function to `rendererPath`, and it's technically possible that some 3rd-party code is importing and using that. But I assume that's ok since it's not part of Gutenberg's documented API.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

alexsanford committed:
Add custom URL query arguments to ServerSideRender

alexsanford committed:
Add wrapper to editor package for ServerSideRender

alexsanford committed:
Add README

alexsanford committed:
Update block library components to use ServerSideRender wrapper

jorgefilipecosta commented on packages/components/src/server-side-render/index.js:
```diff
@@ -20,10 +20,11 @@ import { addQueryArgs } from '@wordpress/url';
 import Spinner from '../spinner';
 
-export function rendererPathWithAttributes( block, attributes = null ) {
+export function rendererPath( block, attributes = null, urlQueryArgs = null ) {
```
Minor: I think for this case we may be able to just do urlQueryArgs = {} on the parameter list, and then ...urlQueryArgs inside addQueryArgs.

alexsanford replied:
Good call. Done.

jorgefilipecosta commented on packages/editor/src/utils/server-side-render/index.js:
```diff
@@ -0,0 +1,15 @@
+	const { getCurrentPostId } = select( 'core/editor' );
+
+	return {
+		urlQueryArgs: {
```
We are ignoring the urlQueryArgs we may be receiving. Maybe we can just check if in urlQueryArgs props this component receives the property post_id is not set and if that's case we add it. urlQueryArgs would default to "{}" so on the most common case we would only add the post_id (equivalent to what we have now).

alexsanford replied:
Good catch! I've updated the implementation for this. It no longer uses `withSelect` but just returns a functional component instead.

jorgefilipecosta commented on packages/editor/src/utils/index.js:
```diff
@@ -2,5 +2,6 @@
  * Internal dependencies
```
We have some prior art in having a component in @wordpress/editor that adds/enhances props to a component in @worpress/components, e.g., https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/panel-color/index.js and we implemented them in src/components maybe we can also add this one there to be consistent.

alexsanford replied:
Done!

jorgefilipecosta reviewed (commented):
This is looking good 👍 Thank you for iterating on this @alexsanford! I left some comments with possible improvements we can do but from the functional point of view, the tests went great.

alexsanford committed:
Simplify addition of urlQueryArgs

alexsanford committed:
Add more explanation to README

alexsanford committed:
Ensure we accept incoming urlQueryArgs rather than blindly overriding them

alexsanford committed:
Move editor ServerSideRender into components subdirectory

alexsanford commented on packages/editor/src/utils/server-side-render/ReadMe.md:
```diff
@@ -0,0 +1,3 @@
+# ServerSideRender
```
Heh. Turns out I had fixed that in a rebase and forgot to push! 🤦‍♂️ 

Updated the text.

alexsanford commented:
@jorgefilipecosta Thanks for the comments! I've addressed them above. Note that since I needed to force push anyway (because of an earlier rebase that didn't get pushed) I went ahead and rebased this onto `master` (no conflicts).

danielbachhuber commented on packages/components/src/server-side-render/index.js:
```diff
@@ -20,10 +20,11 @@ import { addQueryArgs } from '@wordpress/url';
 import Spinner from '../spinner';
 
-export function rendererPathWithAttributes( block, attributes = null ) {
+export function rendererPath( block, attributes = null, urlQueryArgs = {} ) {
```
Given this is an exported function, do we need to keep (or deprecate) the original `rendererPathWithAttributes()`?

danielbachhuber replied:
Phrased another way: maybe we should keep the name as `rendererPathWithAttributes()`, and it simply supports a third `urlQueryArgs` argument.

jorgefilipecosta replied:
Hi @danielbachhuber, the function rendererPathWithAttributes is exported just for testing but is not exported outside of @wordpress/components as we only export the default export outside of the module (the component) https://github.com/WordPress/gutenberg/blob/0c8481cc72c14499dead222a250d96355fdb4f03/packages/components/src/index.js#L51).
No plugin could have used this function so even if we change the name I don't think deprecations are necessary.

danielbachhuber replied:
Ah, makes sense. Thanks for clarifying @jorgefilipecosta

danielbachhuber reviewed (changes requested):
One note to address.

jorgefilipecosta reviewed (approved):
LGTM 👍
